File: src/lib/talentCheck.ts
/**
 * Core Talent Check Evaluation Engine
 * Pure, deterministic functions for skill gap calculation, category aggregation, and candidate readiness scoring.
 */

import {
  SkillCategoryCode,
  SkillGapStatus,
  ReadinessTier,
  ItemizedSkillComparison,
  CategorySkillComparison,
  TalentCheckResponse,
  SKILL_CATEGORY_NAMES,
} from "@/shared/types";
import { DashboardSkill } from "@/lib/companyData";
import { CandidateProfile } from "@/shared/types/candidate";

/**
 * Maps a skill name to one of the 13 canonical RADIX SkillCategoryCodes
 */
export function mapSkillNameToCategoryCode(skillName: string): SkillCategoryCode {
  const name = skillName.toLowerCase();

  if (name.includes("data structure") || name.includes("algorithm") || name.includes("dsa") || name.includes("leetcode")) {
    return "DSA";
  }
  if (name.includes("object-oriented") || name.includes("oop") || name.includes("design pattern") || name.includes("solid")) {
    return "OOD";
  }
  if (name.includes("aptitude") || name.includes("logical reasoning") || name.includes("quantitative") || name.includes("problem solving")) {
    return "APTI";
  }
  if (name.includes("communication") || name.includes("behavioral") || name.includes("soft skill") || name.includes("verbal")) {
    return "COMM";
  }
  if (name.includes("generative ai") || name.includes("genai") || name.includes("ai native") || name.includes("machine learning") || name.includes("deep learning") || name.includes("ai/ml") || name.includes("artificial intelligence") || name.includes("llm")) {
    return "AI";
  }
  if (name.includes("cloud") || name.includes("aws") || name.includes("azure") || name.includes("gcp") || name.includes("devops") || name.includes("docker") || name.includes("kubernetes")) {
    return "CLOUD";
  }
  if (name.includes("sql") || name.includes("database") || name.includes("postgres") || name.includes("mysql") || name.includes("mongodb") || name.includes("redis")) {
    return "SQL";
  }
  if (name.includes("system design") || name.includes("distributed system") || name.includes("architecture") || name.includes("scalability") || name.includes("microservice")) {
    return "SYSD";
  }
  if (name.includes("network") || name.includes("tcp") || name.includes("http") || name.includes("dns") || name.includes("socket")) {
    return "NETW";
  }
  if (name.includes("operating system") || name.includes("linux") || name.includes("kernel") || name.includes("unix") || name.includes("posix") || name.includes("concurrency")) {
    return "OS";
  }
  if (name.includes("software engineering") || name.includes("git") || name.includes("version control") || name.includes("testing") || name.includes("agile") || name.includes("ci/cd")) {
    return "SWE";
  }
  if (name.includes("coding") || name.includes("programming") || name.includes("python") || name.includes("java") || name.includes("c++") || name.includes("javascript") || name.includes("typescript") || name.includes("web development") || name.includes("react")) {
    return "COD";
  }

  return "OTHER";
}

/**
 * Calculates deterministic skill gap:
 * gap = max(0, required_level - candidate_level)
 * A candidate who exceeds the target receives gap = 0 (never negative).
 */
export function calculateSkillGap(
  requiredLevel: number,
  candidateLevel: number
): { gap: number; status: SkillGapStatus } {
  const req = Math.max(1, Math.min(10, Math.round(requiredLevel)));
  const cand = Math.max(1, Math.min(10, Math.round(candidateLevel)));
  const gap = Math.max(0, req - cand);

  let status: SkillGapStatus = "met";
  if (gap > 2) {
    status = "critical_gap";
  } else if (gap > 0) {
    status = "minor_gap";
  }

  return { gap, status };
}

/**
 * Calculates overall candidate readiness score across required competencies.
 * Formula:
 * Readiness = round( (Sum of min(candidate_level, required_level)) / (Sum of required_level) * 100 )
 * Bounded strictly between 0 and 100%.
 */
export function calculateOverallReadiness(
  comparisons: ItemizedSkillComparison[]
): { score: number; tier: ReadinessTier } {
  if (!comparisons || comparisons.length === 0) {
    return { score: 100, tier: "Ready" };
  }

  let totalRequired = 0;
  let totalMet = 0;

  for (const item of comparisons) {
    const req = Math.max(1, Math.min(10, item.required_level));
    const cand = Math.max(1, Math.min(10, item.candidate_level));
    totalRequired += req;
    totalMet += Math.min(cand, req);
  }

  if (totalRequired === 0) {
    return { score: 100, tier: "Ready" };
  }

  const rawScore = (totalMet / totalRequired) * 100;
  const score = Math.max(0, Math.min(100, Math.round(rawScore)));

  let tier: ReadinessTier = "Ready";
  if (score < 60) {
    tier = "Significant Gap";
  } else if (score < 80) {
    tier = "Needs Preparation";
  }

  return { score, tier };
}

/**
 * Identifies top priority skill gaps (where candidate is furthest below target).
 */
export function extractPriorityGaps(
  comparisons: ItemizedSkillComparison[],
  limit = 3
): ItemizedSkillComparison[] {
  return comparisons
    .filter((c) => c.gap > 0)
    .sort((a, b) => {
      // Sort primarily by gap (descending)
      if (b.gap !== a.gap) return b.gap - a.gap;
      // Secondarily by required_level (descending)
      return b.required_level - a.required_level;
    })
    .slice(0, limit);
}

/**
 * Aggregates itemized comparisons into category-level comparisons.
 */
export function aggregateByCategory(
  comparisons: ItemizedSkillComparison[]
): CategorySkillComparison[] {
  const map = new Map<SkillCategoryCode, ItemizedSkillComparison[]>();

  for (const item of comparisons) {
    if (!map.has(item.category_code)) {
      map.set(item.category_code, []);
    }
    map.get(item.category_code)!.push(item);
  }

  const result: CategorySkillComparison[] = [];

  for (const [code, items] of map.entries()) {
    const reqSum = items.reduce((acc, i) => acc + i.required_level, 0);
    const candSum = items.reduce((acc, i) => acc + i.candidate_level, 0);
    const gapSum = items.reduce((acc, i) => acc + i.gap, 0);
    const count = items.length;

    const reqAvg = Number((reqSum / count).toFixed(1));
    const candAvg = Number((candSum / count).toFixed(1));
    const gapAvg = Number((gapSum / count).toFixed(1));

    let status: SkillGapStatus = "met";
    if (gapAvg > 2) {
      status = "critical_gap";
    } else if (gapAvg > 0) {
      status = "minor_gap";
    }

    result.push({
      category_code: code,
      category_name: SKILL_CATEGORY_NAMES[code] || code,
      required_level_avg: reqAvg,
      candidate_level_avg: candAvg,
      gap_avg: gapAvg,
      status,
      skills: items,
    });
  }

  return result.sort((a, b) => b.gap_avg - a.gap_avg);
}

/**
 * Evaluates full Talent Check response from company skills and candidate levels.
 */
export function evaluateTalentCheck(
  candidate: CandidateProfile,
  companyId: number,
  companyName: string,
  companySkills: DashboardSkill[],
  assessmentOverrides?: Record<string, number>
): TalentCheckResponse {
  // Map candidate skills for quick lookup by name or category
  const candidateSkillMap = new Map<string, number>();
  for (const s of candidate.skills || []) {
    candidateSkillMap.set(s.skill_name.toLowerCase(), s.level || 5);
  }

  // Build itemized comparison for each company required skill
  const itemized: ItemizedSkillComparison[] = companySkills.map((cSkill) => {
    const catCode = mapSkillNameToCategoryCode(cSkill.name);

    // Determine candidate level from overrides > candidate profile > default estimate (4)
    let candLevel = 4;
    if (assessmentOverrides && typeof assessmentOverrides[cSkill.name] === "number") {
      candLevel = assessmentOverrides[cSkill.name];
    } else if (candidateSkillMap.has(cSkill.name.toLowerCase())) {
      candLevel = candidateSkillMap.get(cSkill.name.toLowerCase())!;
    } else {
      // Match by category code if skill not directly named
      const matchingSkill = (candidate.skills || []).find((s) => s.category_code === catCode);
      if (matchingSkill && matchingSkill.level) {
        candLevel = matchingSkill.level;
      }
    }

    const { gap, status } = calculateSkillGap(cSkill.score, candLevel);

    return {
      skill_name: cSkill.name,
      category_code: catCode,
      required_level: cSkill.score,
      candidate_level: candLevel,
      gap,
      status,
      evidence: `Assessed Level ${candLevel}/10 vs Company Benchmark ${cSkill.score}/10`,
    };
  });

  const { score: overallReadiness, tier } = calculateOverallReadiness(itemized);
  const priorityGaps = extractPriorityGaps(itemized, 3);
  const categories = aggregateByCategory(itemized);

  const keyStrengths = itemized
    .filter((i) => i.gap === 0 && i.candidate_level >= 7)
    .map((i) => `${i.skill_name} (Level ${i.candidate_level}/10)`);

  const priorityGapDescriptions = priorityGaps.map(
    (g) => `${g.skill_name}: Gap of ${g.gap} ${g.gap === 1 ? "level" : "levels"} (Current L${g.candidate_level} vs Target L${g.required_level})`
  );

  return {
    candidate_id: candidate.id,
    company_id: companyId,
    company_name: companyName,
    overall_readiness_score: overallReadiness,
    readiness_tier: tier,
    category_comparisons: categories,
    key_strengths: keyStrengths.length > 0 ? keyStrengths : ["Baseline foundational competency demonstrated"],
    priority_gaps: priorityGapDescriptions,
    evaluated_at: new Date().toISOString(),
  };
}

File: backend/app/services/talent_check.py
"""
Talent Check Service
Deterministic evaluation engine comparing Candidate Competencies against Company Benchmarks.
"""

from datetime import datetime, timezone
from typing import Dict, List, Optional
from app.models.skill import SkillCategoryCode
from app.models.talent_check import (
    SkillGapStatus,
    ReadinessTier,
    ItemizedSkillComparison,
    CategorySkillComparison,
    TalentCheckRequest,
    TalentCheckResponse,
)

# Standard category names lookup
CATEGORY_NAMES: Dict[SkillCategoryCode, str] = {
    SkillCategoryCode.COD: "Coding",
    SkillCategoryCode.DSA: "Data Structures & Algorithms",
    SkillCategoryCode.OOD: "Object-Oriented Design",
    SkillCategoryCode.APTI: "Aptitude",
    SkillCategoryCode.COMM: "Communication",
    SkillCategoryCode.AI: "Artificial Intelligence",
    SkillCategoryCode.CLOUD: "Cloud",
    SkillCategoryCode.SQL: "SQL",
    SkillCategoryCode.SWE: "Software Engineering",
    SkillCategoryCode.SYSD: "System Design",
    SkillCategoryCode.NETW: "Networking",
    SkillCategoryCode.OS: "Operating Systems",
    SkillCategoryCode.OTHER: "Other",
}

# Standard company benchmark fallback templates for top companies
COMPANY_BENCHMARKS: Dict[int, Dict[str, int]] = {
    # 1: Accenture
    1: {
        "Data Structures & Algorithms": 7,
        "Object-Oriented Programming": 8,
        "SQL & Databases": 7,
        "Cloud Fundamentals (AWS/Azure)": 6,
        "Operating Systems": 6,
        "Computer Networks": 5,
        "Aptitude & Logical Reasoning": 8,
        "Communication & Behavioral": 8,
        "Web Development Basics": 6,
        "System Design (Intro)": 5,
        "Git & Version Control": 6,
        "Generative AI Basics": 6,
    },
    # 2: Google
    2: {
        "Data Structures & Algorithms": 9,
        "System Design & Architecture": 8,
        "Coding & Problem Solving": 9,
        "Operating Systems": 7,
        "Computer Networks": 7,
        "AI Native Engineering": 8,
        "Object-Oriented Programming": 8,
        "SQL & Data Architecture": 7,
        "Cloud Infrastructure": 8,
        "Communication & Leadership": 8,
    },
    # 3: Microsoft
    3: {
        "Data Structures & Algorithms": 8,
        "Object-Oriented Design": 8,
        "Cloud Computing (Azure)": 8,
        "Software Engineering & Testing": 8,
        "System Design": 8,
        "SQL & Databases": 7,
        "Operating Systems": 7,
        "AI & Machine Learning": 7,
        "Communication Skills": 8,
    },
    # 4: Oracle Financial Services Software
    4: {
        "SQL & Relational Databases": 9,
        "Java & Object-Oriented Programming": 8,
        "Data Structures & Algorithms": 7,
        "System Design & Transaction Integrity": 8,
        "Computer Networks & Security": 7,
        "Operating Systems & Linux": 7,
        "Financial Domain Aptitude": 8,
        "Communication Skills": 7,
    },
}

COMPANY_NAMES: Dict[int, str] = {
    1: "Accenture plc",
    2: "Google",
    3: "Microsoft",
    4: "Oracle Financial Services Software",
}


def map_skill_to_category(skill_name: str) -> SkillCategoryCode:
    """Maps skill names to canonical RADIX SkillCategoryCodes."""
    name = skill_name.lower()
    if any(k in name for k in ["data structure", "algorithm", "dsa", "leetcode"]):
        return SkillCategoryCode.DSA
    if any(k in name for k in ["object-oriented", "oop", "design pattern", "solid"]):
        return SkillCategoryCode.OOD
    if any(k in name for k in ["aptitude", "logical reasoning", "quantitative", "puzzle"]):
        return SkillCategoryCode.APTI
    if any(k in name for k in ["communication", "behavioral", "leadership", "verbal"]):
        return SkillCategoryCode.COMM
    if any(k in name for k in ["generative ai", "genai", "ai", "machine learning", "deep learning", "llm"]):
        return SkillCategoryCode.AI
    if any(k in name for k in ["cloud", "aws", "azure", "gcp", "devops", "docker", "kubernetes"]):
        return SkillCategoryCode.CLOUD
    if any(k in name for k in ["sql", "database", "postgres", "mysql", "mongodb", "transaction"]):
        return SkillCategoryCode.SQL
    if any(k in name for k in ["system design", "distributed system", "architecture", "scalability"]):
        return SkillCategoryCode.SYSD
    if any(k in name for k in ["network", "tcp", "http", "dns", "security"]):
        return SkillCategoryCode.NETW
    if any(k in name for k in ["operating system", "linux", "kernel", "unix", "concurrency"]):
        return SkillCategoryCode.OS
    if any(k in name for k in ["software engineering", "git", "testing", "agile", "ci/cd"]):
        return SkillCategoryCode.SWE
    if any(k in name for k in ["coding", "programming", "python", "java", "c++", "web development"]):
        return SkillCategoryCode.COD
    return SkillCategoryCode.OTHER


def calculate_skill_gap(required: int, candidate: int) -> tuple[int, SkillGapStatus]:
    """Calculates skill gap: gap = max(0, required - candidate). Zero if above target."""
    req = max(1, min(10, required))
    cand = max(1, min(10, candidate))
    gap = max(0, req - cand)

    if gap == 0:
        status = SkillGapStatus.MET
    elif gap <= 2:
        status = SkillGapStatus.MINOR_GAP
    else:
        status = SkillGapStatus.CRITICAL_GAP

    return gap, status


def calculate_readiness(comparisons: List[ItemizedSkillComparison]) -> tuple[float, ReadinessTier]:
    """
    Calculates deterministic readiness score.
    Score = round( (Sum of min(candidate, required)) / (Sum of required) * 100 )
    Bounded between 0 and 100%.
    """
    if not comparisons:
        return 100.0, ReadinessTier.READY

    total_required = sum(c.required_level for c in comparisons)
    total_met = sum(min(c.candidate_level, c.required_level) for c in comparisons)

    if total_required == 0:
        return 100.0, ReadinessTier.READY

    score = round((total_met / total_required) * 100, 1)
    score = max(0.0, min(100.0, score))

    if score >= 80.0:
        tier = ReadinessTier.READY
    elif score >= 60.0:
        tier = ReadinessTier.NEEDS_PREPARATION
    else:
        tier = ReadinessTier.SIGNIFICANT_GAP

    return score, tier


def evaluate_talent_check_request(req: TalentCheckRequest) -> TalentCheckResponse:
    """Evaluates candidate against company placement requirements."""
    company_id = req.company_id
    company_name = COMPANY_NAMES.get(company_id, f"Company {company_id}")
    benchmarks = COMPANY_BENCHMARKS.get(company_id, COMPANY_BENCHMARKS[1])
    custom_assessments = req.custom_skill_assessments or {}

    itemized: List[ItemizedSkillComparison] = []

    for skill_name, required_level in benchmarks.items():
        cat_code = map_skill_to_category(skill_name)
        # Check custom override first, else default baseline of 5
        cand_level = custom_assessments.get(skill_name, 5)
        cand_level = max(1, min(10, cand_level))

        gap, status = calculate_skill_gap(required_level, cand_level)

        itemized.append(
            ItemizedSkillComparison(
                skill_name=skill_name,
                category_code=cat_code,
                required_level=required_level,
                candidate_level=cand_level,
                gap=gap,
                status=status,
                evidence=f"Assessed L{cand_level}/10 vs Company Target L{required_level}/10",
            )
        )

    score, tier = calculate_readiness(itemized)

    # Category aggregation
    cat_map: Dict[SkillCategoryCode, List[ItemizedSkillComparison]] = {}
    for item in itemized:
        cat_map.setdefault(item.category_code, []).append(item)

    categories: List[CategorySkillComparison] = []
    for code, items in cat_map.items():
        req_avg = round(sum(i.required_level for i in items) / len(items), 1)
        cand_avg = round(sum(i.candidate_level for i in items) / len(items), 1)
        gap_avg = round(sum(i.gap for i in items) / len(items), 1)

        if gap_avg == 0:
            c_status = SkillGapStatus.MET
        elif gap_avg <= 2:
            c_status = SkillGapStatus.MINOR_GAP
        else:
            c_status = SkillGapStatus.CRITICAL_GAP

        categories.append(
            CategorySkillComparison(
                category_code=code,
                category_name=CATEGORY_NAMES.get(code, code.value),
                required_level_avg=req_avg,
                candidate_level_avg=cand_avg,
                gap_avg=gap_avg,
                status=c_status,
                skills=items,
            )
        )

    categories.sort(key=lambda c: c.gap_avg, reverse=True)

    # Top priority gaps (gap > 0 sorted by gap desc, required desc)
    priority_items = [i for i in itemized if i.gap > 0]
    priority_items.sort(key=lambda i: (i.gap, i.required_level), reverse=True)
    priority_descriptions = [
        f"{p.skill_name}: Gap of {p.gap} {'level' if p.gap == 1 else 'levels'} (Current L{p.candidate_level} vs Target L{p.required_level})"
        for p in priority_items[:3]
    ]

    key_strengths = [
        f"{i.skill_name} (Level {i.candidate_level}/10)"
        for i in itemized
        if i.gap == 0 and i.candidate_level >= 7
    ]
    if not key_strengths:
        key_strengths = ["Baseline foundational competency demonstrated across core subjects"]

    return TalentCheckResponse(
        candidate_id=req.candidate_id,
        company_id=company_id,
        company_name=company_name,
        overall_readiness_score=score,
        readiness_tier=tier,
        category_comparisons=categories,
        key_strengths=key_strengths,
        priority_gaps=priority_descriptions,
        evaluated_at=datetime.now(timezone.utc).isoformat(),
    )
